beacon-chain/rpc/validator/status.go

```diff
@@ -88,6 +88,101 @@ func (vs *Server) MultipleValidatorStatus(
 	}, nil
 }
 
+// CheckDoppelGanger checks if the provided keys are currently active in the network.
+func (vs *Server) CheckDoppelGanger(ctx context.Context, req *ethpb.DoppelGangerRequest) (*ethpb.DoppelGangerResponse, error) {
+	if vs.SyncChecker.Syncing() {
+		return nil, status.Errorf(codes.Unavailable, "Syncing to latest head, not ready to respond")
+	}
+	if req == nil || req.ValidatorRequests == nil || len(req.ValidatorRequests) == 0 {
+		return &ethpb.DoppelGangerResponse{
+			Responses: []*ethpb.DoppelGangerResponse_ValidatorResponse{},
+		}, nil
+	}
+	headState, err := vs.HeadFetcher.HeadState(ctx)
+	if err != nil {
+		return nil, status.Error(codes.Internal, "Could not get head state")
+	}
+	// We walk back from the current head state to the state at the beginning of the previous 2 epochs.
+	// Where S_i , i := 0,1,2. i = 0 would signify the current head state in this epoch.
+	currEpoch := helpers.SlotToEpoch(headState.Slot())
+	previousEpoch, err := currEpoch.SafeSub(1)
+	if err != nil {
+		previousEpoch = currEpoch
+	}
+	olderEpoch, err := previousEpoch.SafeSub(1)
+	if err != nil {
+		olderEpoch = previousEpoch
+	}
+	prevState, err := vs.StateGen.StateBySlot(ctx, params.BeaconConfig().SlotsPerEpoch.Mul(uint64(previousEpoch)))
+	if err != nil {
+		return nil, status.Error(codes.Internal, "Could not get previous state")
+	}
+	olderState, err := vs.StateGen.StateBySlot(ctx, params.BeaconConfig().SlotsPerEpoch.Mul(uint64(olderEpoch)))
+	if err != nil {
+		return nil, status.Error(codes.Internal, "Could not get older state")
+	}
+	resp := &ethpb.DoppelGangerResponse{
+		Responses: []*ethpb.DoppelGangerResponse_ValidatorResponse{},
+	}
+	for _, v := range req.ValidatorRequests {
+		// If the validator's last recorded epoch was
+		// less than 2 epoch ago, this method will not
+		// be able to catch duplicates.
+		if v.Epoch+2 >= currEpoch {
+			resp.Responses = append(resp.Responses,
+				&ethpb.DoppelGangerResponse_ValidatorResponse{
+					PublicKey:       v.PublicKey,
+					DuplicateExists: false,
+				})
+			continue
+		}
```

Comment:
Why is this the case? It isn't clear to the reader why a request within the last 2 epochs will be marked as false or safe to proceed without actually checking.